salt/cache/mysql_cache.py

```diff
@@ -0,0 +1,225 @@
+_RECONNECT_INTERVAL_SEC = 0.050
+
+log = logging.getLogger(__name__)
+client = None
+_mysql_kwargs = None
+_table_name = None
+
+# Module properties
+
+__virtualname__ = 'mysql'
+__func_alias__ = {'ls': 'list'}
+
+
+def run_query(conn, query, retries=3):
+    '''
+    Get a cursor and run a query. Reconnect up to `retries` times if
+    needed.
+    Returns: cursor, affected rows counter
+    Raises: SaltCacheError, AttributeError, MySQLdb.OperationalError
+    '''
+    try:
+        cur = conn.cursor()
+        out = cur.execute(query)
+        return cur, out
+    except (AttributeError, MySQLdb.OperationalError) as e:
+        if retries == 0:
+            raise
+        # reconnect creating new client
+        sleep(_RECONNECT_INTERVAL_SEC)
+        if conn is None:
+            log.debug("mysql_cache: creating db connection")
+        else:
+            log.info("mysql_cache: recreating db connection due to: %r", e)
+        global client
+        client = MySQLdb.connect(**_mysql_kwargs)
+        return run_query(client, query, retries - 1)
+    except Exception as e:
+        if len(query) > 150:
+            query = query[:150] + "<...>"
+        raise SaltCacheError("Error running {0}: {1}".format(query, e))
+
+
+def _create_table():
+    '''
+    Create table if needed
+    '''
+    # Explicitely check if the table already exists as the library logs a
+    # warning on CREATE TABLE
+    query = """SELECT COUNT(TABLE_NAME) FROM information_schema.tables
+        WHERE table_schema = '{0}' AND table_name = '{1}'""".format(
+            _mysql_kwargs['db'],
+            _table_name,
+        )
+    cur, _ = run_query(client, query)
+    r = cur.fetchone()
+    cur.close()
+    if r[0] == 1:
+        return
+
+    query = """CREATE TABLE IF NOT EXISTS {0} (
+      bank CHAR(255),
+      etcd_key CHAR(255),
+      data MEDIUMBLOB,
+      PRIMARY KEY(bank, etcd_key)
+    );""".format(_table_name)
+    log.info("mysql_cache: creating table %s", _table_name)
+    cur, _ = run_query(client, query)
+    cur.close()
+
+
+def __virtual__():
+    '''
+    Confirm that python-mysql package is installed. Setup client and init
+    datastore.
+    '''
+    if not HAS_MYSQL:
+        return (False, "Please install python-mysql package to use mysql data "
+                "cache driver")
+
+    global _mysql_kwargs, _table_name
+    _mysql_kwargs = {
+        'host': __opts__.get('mysql.host', '127.0.0.1'),
+        'user': __opts__.get('mysql.user', None),
+        'passwd': __opts__.get('mysql.password', None),
+        'db': __opts__.get('mysql.database', _DEFAULT_DATABASE_NAME),
+        'port': __opts__.get('mysql.port', 3306),
+        'unix_socket': __opts__.get('mysql.unix_socket', None),
+        'connect_timeout': __opts__.get('mysql.connect_timeout', None),
+        'autocommit': True,
+    }
+    _table_name = __opts__.get('mysql.table_name', _table_name)
+    # TODO: handle SSL connection parameters
+
+    for k, v in _mysql_kwargs.items():
```

Comment:
I don't think we want all this code in `__virtual__`. It will run every time the loader is invoked and that's going to have an across-the-board performance impact, even for those who don't use this cache back-end.